### PR TITLE
fix mypy with `default_factory=list` etc.

### LINKED ISSUE #4457: mypy error on default_factory=list in pydantic 1.10
### Initial Checks

- [X] I have searched GitHub for a duplicate issue and I'm sure this is something new
- [X] I have searched Google & StackOverflow for a solution and couldn't find anything
- [X] I have read and followed [the docs](https://pydantic-docs.helpmanual.io) and still think this is a bug
- [X] I am confident that the issue is with pydantic  (not my code, or another library in the ecosystem like [FastAPI](https://fastapi.tiangolo.com) or [mypy](https://mypy.readthedocs.io/en/stable))

### Description

mypy complains about default_factory = list in pydantic 1.10

But using [] works fine. I think this inconsistency should be fixed
```
$ mypy bug.py
bug.py:6: error: Incompatible types in assignment (expression has type "List[_T]", variable has type "List[str]")
Found 1 error in 1 file (checked 1 source file)
```

Also, I don't think the documentation indicates that pydantic deep-copies mutable default values. I found that out at https://stackoverflow.com/questions/63793662/how-to-give-a-pydantic-list-field-a-default-value

### Example Code

```Python
from typing import List
from pydantic import BaseModel, Field


class Model(BaseModel):
    l1: List[str] = Field(default_factory=list)
    l2: List[str] = Field([])
    l3: List[str] = []
```

### Python, Pydantic & OS Version

```Text
pydantic version: 1.10.1
            pydantic compiled: True
                 install path: /home/mfulmer/miniconda3/envs/rk310/lib/python3.10/site-packages/pydantic
               python version: 3.10.4 (main, Mar 31 2022, 08:41:55) [GCC 7.5.0]
                     platform: Linux-5.4.0-124-generic-x86_64-with-glibc2.31
     optional deps. installed: ['dotenv', 'email-validator', 'typing-extensions']
```

### Affected Components

- [ ] [Compatibility between releases](https://pydantic-docs.helpmanual.io/changelog/)
- [ ] [Data validation/parsing](https://pydantic-docs.helpmanual.io/usage/models/#basic-model-usage)
- [ ] [Data serialization](https://pydantic-docs.helpmanual.io/usage/exporting_models/) - `.dict()` and `.json()`
- [ ] [JSON Schema](https://pydantic-docs.helpmanual.io/usage/schema/)
- [ ] [Dataclasses](https://pydantic-docs.helpmanual.io/usage/dataclasses/)
- [ ] [Model Config](https://pydantic-docs.helpmanual.io/usage/model_config/)
- [ ] [Field Types](https://pydantic-docs.helpmanual.io/usage/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://pydantic-docs.helpmanual.io/usage/validation_decorator/)
- [ ] [Generic Models](https://pydantic-docs.helpmanual.io/usage/models/#generic-models)
- [ ] [Other Model behaviour](https://pydantic-docs.helpmanual.io/usage/models/) - `construct()`, pickling, private attributes, ORM mode
- [ ] [Settings Management](https://pydantic-docs.helpmanual.io/usage/settings/)
- [X] [Plugins](https://pydantic-docs.helpmanual.io/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

### PR DESCRIPTION
**Pull request by samuelcolvin**
## Change Summary

Fix mypy plugin when using bare types like `list` and `dict` as `default_factory`.

## Related issue number

fix #4457

## Checklist

* [x] Unit tests for the changes exist
* [ ] Tests pass on CI and coverage remains at 100%
* [x] `changes/<pull request or issue id>-<github username>.md` file added describing change

### CODE REVIEW

File: tests/mypy/outputs/plugin-fail-strict.txt [hunk 1]
```diff
@@ -36,9 +36,8 @@
 219: error: Property "y" defined in "FrozenModel" is read-only  [misc]
 240: error: Incompatible types in assignment (expression has type "None", variable has type "int")  [assignment]
 241: error: Incompatible types in assignment (expression has type "None", variable has type "int")  [assignment]
-244: error: Incompatible types in assignment (expression has type "Set[_T]", variable has type "str")  [assignment]
+244: error: Incompatible types in assignment (expression has type "Set[Any]", variable has type "str")  [assignment]
```
Comment (klaa97) [inline tests/mypy/outputs/plugin-fail-strict.txt:39]: Does it make sense to add another entry which checks that a function returning (e.g.) `list[int]` cannot be assigned to `list[str]`?

**Review by klaa97 — APPROVED** [review]
Great fix, I think that it filters out completely the builtins (maybe also functions returning a generic, which I think is the easiest option) 🔝 Sorry again for the bug introduced. 🙏

File: tests/mypy/outputs/plugin-fail-strict.txt (hunk 1 above)
Comment (samuelcolvin) [reply tests/mypy/outputs/plugin-fail-strict.txt:39]: > Sorry again for the bug introduced. :pray:

Really don't worry about it, I create far more bugs than anyone.

Good idea, I'll add.

**Comment by samuelcolvin** [comment]
please review.

File: tests/mypy/outputs/plugin-fail-strict.txt (hunk 1 above)
Comment (samuelcolvin) [reply tests/mypy/outputs/plugin-fail-strict.txt:39]: done.

**Review by hramezani — APPROVED** [review]

**Pull request merged**